### PR TITLE
feat: hot server reload & debugging over docker

### LINKED ISSUE #2650: [DevEx] allow for hot reload of backend and auth
## Context

We currently have to build the backend and auth docker containers in order to check whether a feature we are working on is complete. @JoblersTune is already working on implementing a hot reload for the frontend by mounting the source files into the docker container and using the `remix dev` command to run it inside the docker container if the node-env is development. We would like to have similar functionality for backend and auth.

EDIT: I guess we could even extend that to the MASE.

## Todos
- [ ]

### LINKED ISSUE #2672: Implement Separate Docker Compose Configurations for Development and Production Dockerfiles
## Context

To get the best possible developer experience we want to allow for hot module reloading (HMR) in our local playground. Setting this up requires different dependencies and configurations compared to our production environment. This leads to the necessity of having two Dockerfiles: one tailored for development (devDockerfile) and another optimized for production (Dockerfile).

The goal of this task is to create two separate Docker Compose files: one for the development environment that maximizes developer efficiency and ease of use, and another that ensures we can still easily spin up and interact with our production version Dockerfiles.

## Linked issues:
- https://github.com/interledger/rafiki/issues/2650

## Todos
- [ ]

### PR DESCRIPTION
**Pull request by BlairCurrey**
## Changes proposed in this pull request

- adds hot sever reload on save to auth/backend docker services
- exposes debugger with instructions on how to use in most development-environment-agnostic way

## Context

fixes #2650
fixes #2672 

## Checklist

- [x] Related issues linked using `fixes #number`
- [ ] Tests added/updated
- [x] Documentation added
- [x] Make sure that all checks pass
- [ ] Bruno collection updated

### CODE REVIEW

File: localenv/README.md [hunk 1]
```diff
@@ -139,6 +139,28 @@ Note that you have to go through an additional "login" step by providing you IPv
...
+
+- go to `chrome://inspect` in chromium browser
+- Click "Configure" add the IP addresses and ports detailed above
+- start docker containers
+- click "inspect" on the service you want to debug to open the chromium debugger
+
+You can either trigger the debugger by adding `debugger` statements in code and restarting, or by adding breakpoints directly in the chromium debugger after starting the docker containers.
+
+For more ways to connect to these debuggers, see the Node docs for debugging: https://nodejs.org/en/learn/getting-started/debugging
+
```
Comment (BlairCurrey) [inline localenv/README.md:163]: I wanted to provide specific instructions for a basic way to use the debugger despite the fact that everyone is liable to have a different setup (vscode, jetbrains, text editor, etc.). Thus I provided instructions for setting up the chromium debugging client and left a link to the node docs if people want to tune it more for their own setup. 

Of course this does require a chromium browser but it's the best we can do. Note that it's not possible to use safari, firefox, or any non-chromium browser because node uses the v8 engine, and those browsers do not.

File: packages/auth/package.json [hunk 2]
```diff
@@ -18,7 +18,8 @@
     "prepack": "pnpm build",
     "postinstall": "pnpm copy-op-schemas",
     "copy-op-schemas": "cp ./node_modules/@interledger/open-payments/dist/openapi/specs/auth-server.yaml ./node_modules/@interledger/open-payments/dist/openapi/specs/schemas.yaml ./src/openapi/specs/",
-    "copy-files": "cp src/graphql/schema.graphql dist/graphql/ && cp -r ./src/openapi ./dist/"
+    "copy-files": "cp src/graphql/schema.graphql dist/graphql/ && cp -r ./src/openapi ./dist/",
+    "dev": "nodemon --watch src --ext ts --exec \"tsc && pnpm copy-files && node --inspect=0.0.0.0:9229 dist/index.js\""
```
Comment (BlairCurrey) [inline packages/auth/package.json:22]: This is run from the dev dockerfile. In plain english, it is:

Copying over the gql/openapi schema files, rebuilding, and starting the sever with the debugger. Then on the docker container's src file change, it is repeating the copy/rebuild/server start. The docker container's src file is a bind mount of the host machine's src file. Thus saving changes triggers the server restart.

File: localenv/cloud-nine-wallet/docker-compose.yml [hunk 3]
```diff
@@ -110,7 +122,7 @@ services:
     image: rafiki-frontend
     build:
       context: ../..
-      dockerfile: ./packages/frontend/devDockerfile
+      dockerfile: ./packages/frontend/Dockerfile.dev
```
Comment (BlairCurrey) [inline localenv/cloud-nine-wallet/docker-compose.yml:125]: went ahead and renamed this for consistency. I went with Dockerfile.dev because that seemed to be a more standard convention from examples I saw a variety of places. VSCode also recognizes its a dockerfile from name alone, whereas it did not with `devDockerfile`.

File: localenv/README.md (hunk 1 above)
Comment (BlairCurrey) [reply localenv/README.md:163]: Please let me know if you can follow these instructions to start using the debugger successfully.

**Comment by BlairCurrey** [comment]
It's not a factor in these changes, but I noticed we have docker-compose in our services, like [rafiki/packages/backend/docker-compose.yml](https://github.com/interledger/rafiki/blob/main/packages/backend/docker-compose.yml). Why?

As far as I can tell they are unused. Do we need them? Should I get rid of them?

**Comment by BlairCurrey** [comment]
I considered adding a CI step for checking that frontend starts OK with the production image. This is not trivial to setup since frontend depends on quite a few things (kratos, backend db, etc.). It would also be completely redundant if we
 add any sort of frontend e2e test as part of this issue: https://github.com/interledger/rafiki/issues/2627

**Comment by mkurapov** [comment]
> using ts-node instead of explicit build step. Never really liked the idea but its the more common use-case with nodemon. Tried it briefly but ran into some typescript errors and didn't pursue it further.

@BlairCurrey curious to hear your thoughts on debugging the complied JS files or the TS files directly.

I tried using `ts-node-dev` in this [commit](https://github.com/interledger/rafiki/commit/251defe80db6b0b4661ac86e8b2f06a72071cc08), and it ended up working quite well with the following launch.json for the cloud-nine-backend debugging (with the remote root pointing to the volume you exposed:

```json
    {
      "name": "Attach to docker (cloud-nine-backend)",
      "type": "node",
      "request": "attach",
      "port": 9229,
      "address": "localhost",
      "localRoot": "${workspaceFolder}",
      "remoteRoot": "/home/rafiki/",
      "restart": true
    },
```

Let me know what feels better/what issues you had when trying TS, but it was quite nice debugging the files we end up writing.

**Comment by BlairCurrey** [comment]
> > using ts-node instead of explicit build step. Never really liked the idea but its the more common use-case with nodemon. Tried it briefly but ran into some typescript errors and didn't pursue it further.
> 
> @BlairCurrey curious to hear your thoughts on debugging the complied JS files or the TS files directly.
> 
> I tried using `ts-node-dev` in this [commit](https://github.com/interledger/rafiki/commit/251defe80db6b0b4661ac86e8b2f06a72071cc08), and it ended up working quite well with the following VS code launch.json for debugging cloud-nine-backend (with the remote root pointing to the volume you expose):
> 
> ```json
>     {
>       "name": "Attach to docker (cloud-nine-backend)",
>       "type": "node",
>       "request": "attach",
>       "port": 9229,
>       "address": "localhost",
>       "localRoot": "${workspaceFolder}",
>       "remoteRoot": "/home/rafiki/",
>       "restart": true
>     },
> ```
> 
> Let me know what feels better/what issues you had when trying TS, but it was quite nice debugging the files we end up writing.

Thanks for sharing this - seems promising. I like that we dont have to copy files every chance (since they arent being wiped out by building).

File: localenv/README.md [hunk 4]
```diff
@@ -139,6 +139,28 @@ Note that you have to go through an additional "login" step by providing you IPv
 
 After stopping the script it is necessary to clear the environment using the command described in [Shutting down](#Shutting-down). This is necessary as on a new run of the scripts (with autopeering or not) the wallet address url will differ.
 
+### Debugging
```
Comment (mkurapov) [inline localenv/README.md:142]: Maybe we can add a section with the launch.json configuration for VSCode as well?

Also, did you want to keep it just in this readme, or should we also add it to the documentation site?

File: localenv/mock-account-servicing-entity/app/entry.server.tsx [hunk 5]
```diff
@@ -21,7 +21,7 @@ async function callWithRetry(fn: () => any, depth = 0): Promise<void> {
   try {
     return await fn()
   } catch (e) {
-    if (depth > 7) {
+    if (depth > 10) {
```
[the remark this answers is not in the document]
Comment (mkurapov) [reply localenv/mock-account-servicing-entity/app/entry.server.tsx:24]: Do we still need to change this with ts-node-dev?

[the remark this answers is not in the document]
Comment (BlairCurrey) [reply localenv/mock-account-servicing-entity/app/entry.server.tsx]: no we dont, changed back to 7

File: localenv/README.md (hunk 4 above)
Comment (BlairCurrey) [reply localenv/README.md:142]: added to the documentation site: https://github.com/interledger/rafiki/pull/2718/commits/02a5e6b780037021ec2fe83ffab4c4fe4191e2d3

File: localenv/README.md [hunk 6]
```diff
@@ -139,6 +139,47 @@ Note that you have to go through an additional "login" step by providing you IPv
...
+
+Debuggers for the services are exposed on the following ports:
+
+| IP and Port    | Services                |
+| -------------- | ----------------------- |
+| 127.0.0.1:9229 | Cloud Nine Backend      |
```
Comment (mkurapov) [inline localenv/README.md:148]:
```suggestion
| 127.0.0.1:9229 | Cloud Nine Wallet Backend      |
```
ditto down below

File: localenv/README.md [hunk 7]
```diff
@@ -139,6 +139,47 @@ Note that you have to go through an additional "login" step by providing you IPv
...
+- go to `chrome://inspect`
+- Click "Configure" and add the IP addresses and ports detailed above
+- start docker containers
+- click "inspect" on the service you want to debug to open the chromium debugger
+
+You can either trigger the debugger by adding `debugger` statements in code and restarting, or by adding breakpoints directly in the chromium debugger after starting the docker containers.
+
+#### With Vscode:
+
+For debugging with Vscode, you can add this configuration to your `.vscode/launch.json`):
```
Comment (mkurapov) [inline localenv/README.md:164]:
```suggestion
#### With VS Code:

For debugging with VS Code, you can add this configuration to your `.vscode/launch.json`):
```
💅

File: packages/auth/package.json [hunk 8]
```diff
@@ -66,6 +68,7 @@
     "jest-openapi": "^0.14.2",
     "nock": "^13.5.4",
     "node-mocks-http": "^1.14.1",
+    "nodemon": "^3.1.0",
```
Comment (mkurapov) [inline packages/auth/package.json:71]:
```suggestion
```

**Review by mkurapov — COMMENTED** [review]
v minor comments

File: packages/auth/package.json (hunk 8 above)
Comment (BlairCurrey) [reply packages/auth/package.json]: good catch, changed locally and updated the lockfile

**Review by mkurapov — APPROVED** [review]
